### PR TITLE
Ks biasing

### PR DESCRIPTION
**Pull request by kmdeck**
## Purpose 
Create and analyze biased samples for the KS system

To run, from kurimotosivashinksy/

```
julia --project training.jl Experiment_all_data.toml
julia --project generate_samples.jl Experiment_all_data.toml k_bias n_avg
julia --project analysis.jl Experiment_all_data.toml k_bias n_avg
```

where `k_bias` is the k to use in biasing, and `n_avg` is defined such that we are biasing the samples towards larger values in a region of size `n_avg^2` pixels. if `k_bias = 0`, `n_avg` does not matter, but it should not be <1.

If you want to check the behavior of the model, run in this way - this gives you the autocorrelation time, qq, image, spatial mean, loss curves, etc plots. all the things we use to assess how the base model is performing.

A follow up PR will make these plots publication ready and create the plots we want to show improvement on return times. We'll need an additional script if we want to plot multiple values of `k` in one plot - these scripts only read one set of samples.

## Examples
using Experiment_all_data.toml, k_bias = 0.0, and n_avg = 1:
train_images_ch1
gen_images_ch1
spatial_mean_distribution
qq_plot
event_probability_16
autocorr

### CODE REVIEW

File: examples/kuramotosivashinksy/analysis.jl [hunk 1]
```diff
@@ -8,21 +8,20 @@ using Plots
...
 
 using CliMAgen
 package_dir = pkgdir(CliMAgen)
 include(joinpath(package_dir,"examples/utils_data.jl"))
 include(joinpath(package_dir,"examples/utils_analysis.jl"))
-include(joinpath(package_dir,"examples/utils_etl.jl"))
-
-function run_analysis(params; FT=Float32, logger=nothing)
 
+function run_analysis(params; FT=Float32, k_bias=0.0f0, n_avg=1)
```
Comment (kmdeck) [inline examples/kuramotosivashinksy/analysis.jl:18]: this script now never makes samples, so `make_samples` is not longer in the experiment toml.

File: examples/kuramotosivashinksy/generate_samples.jl [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    else
+        device = Flux.cpu
+        @info "Sampling on CPU"
+    end
+
+    # set up bias for space-time mean
+    indicator = zeros(FT, n_pixels, n_time, inchannels)
+    midx = Int(div(n_pixels, 2))
+    midy = Int(div(n_time, 2))
+    indicator[midx-n_avg+1:midx, midy-n_avg+1:midy, :] .= 1
```
Comment (kmdeck) [inline examples/kuramotosivashinksy/generate_samples.jl:50]: of note

File: examples/kuramotosivashinksy/generate_samples.jl [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+    xtrain = cat([x for x in dl]..., dims=4);
+    Z = mean(exp.(k_bias .* A(xtrain;indicator = cpu(indicator)))[:])
+    # Compute the likelihood ratio of the samples
+    lr = Z.*exp.(-k_bias .*A(all_samples; indicator = cpu(indicator)))[:]
+  
+    samplesdir = joinpath(savedir, "bias_$(FT(k_bias))_n_avg_$(n_avg)_shift_$shift")
+    !ispath(samplesdir) && mkpath(samplesdir)
+    hdf5_path=joinpath(samplesdir, samples_file)
+    fid = HDF5.h5open(hdf5_path, "w")
+    fid["generated_samples"] = all_samples
```
Comment (kmdeck) [inline examples/kuramotosivashinksy/generate_samples.jl:99]: the idea is that this saves everything we need to compute statistics later. Previous we computed the LR in analysis.jl, but then we need to specify `A` there as well, which could lead to a mismatch.

**Review by bischtob — APPROVED** [review]
I like the new way of separating sampling from analysis. I am fine with this :)!

**Pull request merged**